File: src/main/java/work/lclpnet/ap2/api/game/team/TeamManager.java
package work.lclpnet.ap2.api.game.team;

import net.minecraft.server.network.ServerPlayerEntity;
import work.lclpnet.ap2.api.base.Participants;

import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

public interface TeamManager {

    Set<Team> getTeams();

    Optional<net.minecraft.scoreboard.Team> getMinecraftTeam(TeamKey key);

    Optional<Team> getTeam(TeamKey key);

    Optional<Team> getTeam(ServerPlayerEntity player);

    void partitionIntoTeams(Set<ServerPlayerEntity> players, Set<TeamKey> teams);

    boolean isParticipating(TeamKey key);

    void setTeamEliminated(Team team);

    void bind(TeamEliminatedListener listener);

    default void partitionIntoTeams(Participants participants, Set<TeamKey> teams) {
        partitionIntoTeams(participants.getAsSet(), teams);
    }

    default Set<net.minecraft.scoreboard.Team> getMinecraftTeams() {
        return getTeams().stream()
                .map(Team::getKey)
                .map(this::getMinecraftTeam)
                .flatMap(Optional::stream)
                .collect(Collectors.toUnmodifiableSet());
    }

    default boolean isParticipating(Team team) {
        return isParticipating(team.getKey());
    }

    default boolean isParticipating(ServerPlayerEntity player) {
        return getTeam(player).map(Team::getKey).map(this::isParticipating).orElse(false);
    }

    default Set<Team> getParticipatingTeams() {
        return getTeams().stream()
                .filter(this::isParticipating)
                .collect(Collectors.toUnmodifiableSet());
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/team/TeamManager.kt
package work.lclpnet.ap2.game.team

import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.scores.PlayerTeam
import java.util.*

interface TeamManager {

    val teams: Set<Team>

    fun getMinecraftTeam(key: TeamKey): PlayerTeam?

    fun getTeam(key: TeamKey): Team?

    fun getTeam(uuid: UUID): Team?

    fun partitionIntoTeams(players: Iterable<ServerPlayer>, teams: Iterable<TeamKey>)

    fun isParticipating(key: TeamKey): Boolean

    fun setTeamEliminated(team: Team)

    fun bind(listener: TeamEliminatedListener?)

    /**
     * Whether to use team color codes.
     * This enables colored player name tags above the player model.
     * However, only a limited number of colors codes exist.
     * Set this to true, if all your team rgb colors have matching color codes (Formatting).
     */
    fun setUseColorCodes(useColorCodes: Boolean)

    fun registerTeam(key: TeamKey): Team

    fun joinTeam(player: ServerPlayer, team: Team)

    val minecraftTeams: Set<PlayerTeam>
        get() = teams
            .map { it.key }
            .mapNotNull { getMinecraftTeam(it) }
            .toSet()

    val participatingTeams: Set<Team>
        get() = teams
            .filter { isParticipating(it) }
            .toSet()

    fun getTeam(player: ServerPlayer): Team? =
        getTeam(player.getUUID())

    fun isParticipating(team: Team): Boolean =
        isParticipating(team.key)

    fun isParticipating(player: ServerPlayer): Boolean {
        val team = getTeam(player) ?: return false

        return isParticipating(team)
    }

    fun isTeamMember(player: ServerPlayer, team: Team): Boolean {
        val playerTeam = getTeam(player)

        if (playerTeam != null) {
            return playerTeam == team
        }

        return false
    }

    fun areTeamMates(first: ServerPlayer, second: ServerPlayer): Boolean {
        val team = getTeam(first) ?: return false

        return isTeamMember(second, team)
    }

    fun getTeam(keyable: TeamKeyable): Team? =
        getTeam(keyable.key)
}